core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala

```diff
@@ -76,6 +79,17 @@ class ResourceProfileBuilder() {
     this
   }
 
+  /**
+   * Specify target resource profile type.
+   * @param isTaskResourceProfile will create a [[TaskResourceProfile]] if true, otherwise a
+   *                              normal [[ResourceProfile]]
+   * @return This ResourceProfileBuilder
+   */
+  def taskOnly(isTaskResourceProfile: Boolean = true): this.type = {
```

Comment:
Why do we still need `isTaskResourceProfile` since the function is already named `taskOnly()`?

Reply:
@tgravescs This introduces the new API but doesn't affect the existing ones. Are you good with it?

Reply:
> Why do we still need `isTaskResourceProfile` since the function is already named `taskOnly()`?

Just want to provide a method that user can override the property `isTaskResourceProfile ` to reuse the resource profile builder to create normal `ResourceProfile`, since there are existing methods to override/cleanup resource requests: [clearExecutorResourceRequests](https://github.com/apache/spark/blob/v3.3.0/core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala#L79)

Do you think is it necessary? or any other suggestions?

Reply:
I don't think it's necessary. To me, the name `taskOnly()` implies `isTaskResourceProfile=true`. So setting `isTaskResourceProfile=false` is very wired.

Reply:
Thanks. Then I will remove it.

Reply:
agree, seems unnecessary at this point, but seems you already removed.

Reply:
Hey @tgravescs , there are some difference here I think.
In this PR, we add a new `TaskResourceProfile`, which is limited to some scenarios like dynamic allocation off and it will reuse executors. So that we may want user to explicitly specify it's a `TaskResourceProfile` to avoid misunderstanding.

With dynamic allocation enabled, user can still create a `ResourceProfile` without specifying executor requirements, and request new executors for this rp.

Does this make sense?

Reply:
> If user hasn't specified executor requirements it can just be task only requirements.

Sounds good to me. And we can make `TaskResourceProfile` internally for now.

Reply:
> With dynamic allocation enabled, user can still create a ResourceProfile without specifying executor requirements, and request new executors for this rp.

Is the cluster manager able to launch the new executor If there're no executor requirements (e.g., no memory, no CPU) specified? @ivoson

Reply:
> Is the cluster manager able to launch the new executor If there're no executor requirements (e.g., no memory, no CPU) specified? @ivoson

Yes, cluster manager will launch new executors based on default memory/cpu.

Reply:
>In this PR, we add a new TaskResourceProfile, which is limited to some scenarios like dynamic allocation off and it will reuse executors. So that we may want user to explicitly specify it's a TaskResourceProfile to avoid misunderstanding.

Why does this the user have to explicitly set it.. isn't it implicit if they don't specify executor resources and they call build()?  I guess you could argue you could perhaps give them a better error message if that isn't really what they intended, but that likely still becomes an issue with dynamic allocation support.  I'm fine with TaskResourceProfile being public as a shortcut after building, especially if it makes the schedule code cleaner to check instance type.

The only difference I can see between this and with dynamic allocation is with dynamic allocation if you don't specify the executor resources it defaults to use the initial resources but it gets a new executor based on the current implementation behavior. if we support reuse executors with dynamic allocation that restriction goes away and I would expect it to find any executor that would fit - or like the linked issue we would specify some sort of reuse policy that user could indicate which type of executors to reuse.   Is there some other use case you had in mind?

Please don't force push the code, it makes reviewing what changed much harder.

Reply:
> Please don't force push the code, it makes reviewing what changed much harder.
Thanks. Will avoid force push in future.

> but that likely still becomes an issue with dynamic allocation support.
Yes, agree. My previous concern is that we try to limit `TaskResourceProfile` to dynamic allocation off, exception will be thrown  if user build a `TaskResourceProfile` when dynamic allocation enabled. This will introduce behavior change.
If we don't have the limit, then we need to think about the policy for scheduling tasks with `TaskResourceProfile` when dynamic allocation is enabled.  

cc @Ngone51 @tgravescs What do you think? Shall we support dynamic allocation as well here?